sol-transactions/convo.go

```diff
@@ -0,0 +1,322 @@
+	if err := c.validate(); err != nil {
+		return loop.Quit(err)
+	}
+	return c.state.NextStep()
+}
+
+func (p *Project) NextStep() (out loop.Cmd) {
+	if p.Name == "" {
+		return cmd(codegen.AskProjectName{})
+	}
+
+	if !p.InitialBlockSet {
+		return cmd(codegen.AskInitialStartBlockType{})
+	}
+
+	if p.ProgramId == "" {
+		return cmd(AskProgramId{})
+	}
+
+	if !p.generatedCodeCompleted {
+		return cmd(codegen.RunGenerate{})
+	}
+
+	// Remote build part removed for the moment
+	// if !p.confirmDoCompile && !p.confirmDownloadOnly {
+	// 	return cmd(codegen.AskConfirmCompile{})
+	// }
+
+	return cmd(codegen.RunBuild{})
+}
+
+func (c *Convo) Update(msg loop.Msg) loop.Cmd {
+	if os.Getenv("SUBSTREAMS_DEV_DEBUG_CONVERSATION") == "true" {
+		fmt.Printf("convo Update message: %T %#v\n-> state: %#v\n\n", msg, msg, c.state)
+	}
+
+	switch msg := msg.(type) {
+	case codegen.MsgStart:
+		var msgCmd loop.Cmd
+		if msg.Hydrate != nil {
+			if err := json.Unmarshal([]byte(msg.Hydrate.SavedState), &c.state); err != nil {
+				return loop.Quit(fmt.Errorf(`something went wrong, here's an error message to share with our devs (%s); we've notified them already`, err))
+			}
+
+			msgCmd = c.msg().Message("Ok, I reloaded your state.").Cmd()
+		} else {
+			msgCmd = c.msg().Message("Ok, let's start a new package.").Cmd()
+		}
+		return loop.Seq(msgCmd, c.NextStep())
+
+	case codegen.AskProjectName:
+		return c.action(codegen.InputProjectName{}).
+			TextInput(codegen.InputProjectNameTextInput(), "Submit").
+			Description(codegen.InputProjectNameDescription()).
+			DefaultValue("my_project").
+			Validation(codegen.InputProjectNameRegex(), codegen.InputProjectNameValidation()).
+			Cmd()
+
+	case codegen.InputProjectName:
+		c.state.Name = msg.Value
+		return c.NextStep()
+
+	case codegen.AskInitialStartBlockType:
+		return c.action(codegen.InputAskInitialStartBlockType{}).
+			TextInput(codegen.InputAskInitialStartBlockTypeTextInput(), "Submit").
+			DefaultValue("0").
+			Validation(codegen.InputAskInitialStartBlockTypeRegex(), codegen.InputAskInitialStartBlockTypeValidation()).
+			Cmd()
+
+	case codegen.InputAskInitialStartBlockType:
+		initialBlock, err := strconv.ParseUint(msg.Value, 10, 64)
+		if err != nil {
+			return loop.Quit(fmt.Errorf("invalid start block input value %q, expected a number", msg.Value))
+		}
+
+		c.state.InitialBlock = initialBlock
+		c.state.InitialBlockSet = true
+		return c.NextStep()
+
+	case AskProgramId:
+		return c.action(InputProgramId{}).
+			TextInput(fmt.Sprintf("Filter the transactions based on one or several Program IDs.\n\nSupported operators are: logical or '||', logical and '&&' and parenthesis: '()'. \n\nExample: to only consume TRANSACTIONS containing Token or ComputeBudget instructions: 'program:TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA || program:ComputeBudget111111111111111111111111111111'. \n\nTransactions containing 'Vote111111111111111111111111111111111111111' instructions are always excluded."), "Submit").
```

Comment:
This language is common across all of our modules.

@enoldev could we have a well formatted documentation page that explains all the formats and syntax of that language.. and then link from here?

Reply:
The syntax is specific to every foundational module, do you mean a document with instructions for every foundational module?

Reply:
@enoldev , no, the syntax is generic and used in any module using a filter.

the parenthesis, the && and the || ...

Reply:
But since this is on a Prompt, it makes sense to print it all out here, so I'm good with this code.